phoenix-core/src/main/java/org/apache/phoenix/schema/types/PJson.java

```diff
@@ -0,0 +1,223 @@
+
+    }
+
+    @Override
+    public boolean isSizeCompatible(ImmutableBytesWritable ptr, Object value,
+            @SuppressWarnings("rawtypes") PDataType srcType, Integer maxLength, Integer scale,
+            Integer desiredMaxLength, Integer desiredScale) {
+        return PVarchar.INSTANCE.isSizeCompatible(ptr, value, srcType, maxLength, scale,
+            desiredMaxLength, desiredScale);
+    }
+
+    @Override
+    public boolean isFixedWidth() {
+        return false;
+    }
+
+    @Override
+    public int estimateByteSize(Object o) {
+        PhoenixJson phoenixJson = (PhoenixJson) o;
+        return phoenixJson.estimateByteSize();
+    }
+
+    @Override
+    public Integer getByteSize() {
+        return null;
+    }
+
+    @Override
+    public int compareTo(Object lhs, Object rhs, @SuppressWarnings("rawtypes") PDataType rhsType) {
+        if (PJson.INSTANCE != rhsType) {
+            throwConstraintViolationException(rhsType, this);
+        }
+        PhoenixJson phoenixJsonLHS = (PhoenixJson) lhs;
+        PhoenixJson phoenixJsonRHS = (PhoenixJson) rhs;
+        return phoenixJsonLHS.compareTo(phoenixJsonRHS);
+    }
+
+    @Override
+    public Object toObject(String value) {
+        return getPhoenixJson(value);
+    }
+
+    @Override
+    public boolean isBytesComparableWith(@SuppressWarnings("rawtypes") PDataType otherType) {
+        return otherType == PJson.INSTANCE || otherType == PVarchar.INSTANCE;
+    }
+
+    @Override
+    public String toStringLiteral(Object o, Format formatter) {
+        if (o == null) {
+            return StringUtil.EMPTY_STRING;
+        }
+        PhoenixJson phoenixJson = (PhoenixJson) o;
+        return PVarchar.INSTANCE.toStringLiteral(phoenixJson.toString(), formatter);
+    }
+
+    @Override
+    public Object getSampleValue(Integer maxLength, Integer arrayLength) {
+        Preconditions.checkArgument(maxLength == null || maxLength >= 0);
+
+        char[] key = new char[4];
+        char[] value = new char[4];
+        int length = maxLength != null ? maxLength : 1;
+        if (length > (key.length + value.length)) {
+            key = new char[length + 2];
+            value = new char[length - key.length];
+        }
+        int j = 1;
+        key[0] = '"';
+        key[j++] = 'k';
+        for (int i = 2; i < key.length - 1; i++) {
+            key[j++] = (char) ('0' + RANDOM.get().nextInt(Byte.MAX_VALUE) % 10);
+        }
+        key[j] = '"';
+
+        int k = 1;
+        value[0] = '"';
+        value[k++] = 'v';
+        for (int i = 2; i < value.length - 1; i++) {
+            value[k++] = (char) ('0' + RANDOM.get().nextInt(Byte.MAX_VALUE) % 10);
+        }
+        value[k] = '"';
+        StringBuilder sbr = new StringBuilder();
+        sbr.append("{").append(key).append(":").append(value).append("}");
+
+        byte[] bytes = Bytes.toBytes(sbr.toString());
+        return getPhoenixJson(bytes, 0, bytes.length);
```

Comment:
Get rid of this method: getPhoenixJson(byte[], offset, length). Instead, just have getPhoenixJson(String), since you're relying on having a String. Converting from a String -> byte[] -> String which is a waste.